bosond.cpp

```diff
@@ -99,125 +106,133 @@ int main(int argc, char** argv) {
     format.fmt.pix.height = height;
 
     // request desired FORMAT
-    if(ioctl(fd, VIDIOC_S_FMT, &format) < 0){
+    if (ioctl(fd, VIDIOC_S_FMT, &format) < 0) {
         perror("VIDIOC_S_FMT");
         exit(1);
     }
 
-    // we need to inform the device about buffers to use.
-    // and we need to allocate them.
-    // we’ll use a single buffer, and map our memory using mmap.
-    // All this information is sent using the VIDIOC_REQBUFS call and a
-    // v4l2_requestbuffers structure:
+    // Allocate 2 mmap buffers for retrieving video frames.
     struct v4l2_requestbuffers bufrequest;
     bufrequest.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
     bufrequest.memory = V4L2_MEMORY_MMAP;
-    bufrequest.count = 1;   // we are asking for one buffer
-
-    if(ioctl(fd, VIDIOC_REQBUFS, &bufrequest) < 0){
+    bufrequest.count = num_buffers;
+    if (ioctl(fd, VIDIOC_REQBUFS, &bufrequest) < 0) {
         perror("VIDIOC_REQBUFS");
         exit(1);
     }
 
-    // Now that the device knows how to provide its data,
-    // we need to ask it about the amount of memory it needs,
-    // and allocate it. This information is retrieved using the VIDIOC_QUERYBUF call,
-    // and its v4l2_buffer structure.
-    struct v4l2_buffer bufferinfo;
-    memset(&bufferinfo, 0, sizeof(bufferinfo));
-
-    bufferinfo.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
-    bufferinfo.memory = V4L2_MEMORY_MMAP;
-    bufferinfo.index = 0;
-
-    if (ioctl(fd, VIDIOC_QUERYBUF, &bufferinfo) < 0) {
-        perror("VIDIOC_QUERYBUF");
-        exit(1);
-    }
+    // Now find out about the buffers that were created and map them.
+    struct v4l2_buffer bufferinfo[2];
+    void *buffer[num_buffers];
+    for (int i = 0; i < num_buffers; i++) {
+        memset(&bufferinfo[i], 0, sizeof(struct v4l2_buffer));
+        bufferinfo[i].type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+        bufferinfo[i].memory = V4L2_MEMORY_MMAP;
+        bufferinfo[i].index = i;
+        if (ioctl(fd, VIDIOC_QUERYBUF, &bufferinfo[i]) < 0) {
+            perror("VIDIOC_QUERYBUF");
+            exit(1);
+        }
 
-    void *buffer_start = mmap(NULL, bufferinfo.length, PROT_READ | PROT_WRITE, MAP_SHARED, fd, bufferinfo.m.offset);
-    if (buffer_start == MAP_FAILED) {
-        perror("mmap");
-        exit(1);
+        buffer[i] = mmap(NULL, bufferinfo[0].length, PROT_READ | PROT_WRITE, MAP_SHARED, fd, bufferinfo[i].m.offset);
```

Comment:
reference the length of the buffer at position i not 0 `buferinfo[i].length`

Reply:
Another good catch. Each buffer should be the same length but it's better to be clearer.